cpp/src/arrow/dataset/filter.h

```diff
@@ -305,6 +316,36 @@ class ARROW_DS_EXPORT NotExpression final
   Result<std::shared_ptr<DataType>> Validate(const Schema& schema) const override;
 };
 
+/// Explicitly cast an expression to a different type
+class ARROW_DS_EXPORT CastExpression final
+    : public ExpressionImpl<UnaryExpression, CastExpression, ExpressionType::CAST> {
+ public:
+  CastExpression(std::shared_ptr<Expression> operand, std::shared_ptr<DataType> to,
+                 compute::CastOptions options)
+      : ExpressionImpl(std::move(operand)),
+        to_(std::move(to)),
+        options_(std::move(options)) {}
+
+  CastExpression(std::shared_ptr<Expression> operand, std::shared_ptr<Expression> like,
+                 compute::CastOptions options)
+      : ExpressionImpl(std::move(operand)),
+        like_(std::move(like)),
+        options_(std::move(options)) {}
+
+  std::string ToString() const override;
+
+  std::shared_ptr<Expression> Assume(const Expression& given) const override;
+
+  Result<std::shared_ptr<DataType>> Validate(const Schema& schema) const override;
+
+  const compute::CastOptions& options() const { return options_; }
+
+ private:
+  std::shared_ptr<DataType> to_;
+  std::shared_ptr<Expression> like_;
```

Comment:
Could we either use a variant or  subclasses (CastTo, CastLike), your choice.

Reply:
I'll use a variant